core/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java

```diff
@@ -177,19 +183,30 @@ public void renameAllTempFiles() throws IOException {
     }
 
     /**
-     * Closes the current recovery target and returns a
-     * clone to reset the ongoing recovery.
-     * Note: the returned target must be canceled, failed or finished
-     * in order to release all it's reference.
+     * Closes the current recovery target and waits up to a certain timeout for resources to be freed
      */
-    RecoveryTarget resetRecovery() throws IOException {
+    void resetRecovery() throws InterruptedException {
         ensureRefCount();
         if (finished.compareAndSet(false, true)) {
+            logger.debug("reset of recovery with shard {} and id [{}]", shardId, recoveryId);
             // release the initial reference. recovery files will be cleaned as soon as ref count goes to zero, potentially now
             decRef();
+            closedLatch.await();
+            RecoveryState.Stage stage = indexShard.recoveryState().getStage();
+            if (indexShard.recoveryState().getPrimary() && (stage == RecoveryState.Stage.FINALIZE || stage == RecoveryState.Stage.DONE)) {
```

Comment:
can we something to the PR description about this? it's subtle but important imo.